app/src/main/java/com/example/habitsmasher/MapDialog.java

```diff
@@ -0,0 +1,210 @@
+ * example/mapdemo/RawMapViewDemoActivity.java
+ *
+ * Requesting a location came from following the tutorial here:
+ * https://developer.android.com/training/location/retrieve-current
+ *
+ */
+public class MapDialog extends DialogFragment implements OnMapReadyCallback {
+    private static final String MAPVIEW_BUNDLE_KEY = "MapViewBundleKey";
+    private static final String TAG = "MAP DIALOG";
+
+    private String _selectedLocation;
+    private MapView _mapView;
+    private Button _confirmButton;
+    private Button _cancelButton;
+    private HabitEventDialog _habitEventDialog;
+    private FusedLocationProviderClient _fusedLocationClient;
+    private Bundle _mapViewBundle;
+    private boolean _mapCreated = false;
+
+    public MapDialog(String location) {
+        _selectedLocation = location;
+    }
+
+    @Nullable
+    @Override
+    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
+        View view = inflater.inflate(R.layout.map_dialog, container, false);
+
+
+        _mapViewBundle = null;
+        if (savedInstanceState != null) {
+            _mapViewBundle = savedInstanceState.getBundle(MAPVIEW_BUNDLE_KEY);
+        }
+        _mapView = (MapView) view.findViewById(R.id.edit_map);
+        _cancelButton = view.findViewById(R.id.cancel_map);
+        _confirmButton = view.findViewById(R.id.confirm_map);
+        _cancelButton.setOnClickListener(new ClickListenerForCancel(getDialog(), TAG));
+        // send location back to habit event
+        _confirmButton.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View view){
+                _habitEventDialog.selectLocation(_selectedLocation);
+                getDialog().dismiss();
+            }
+        });
+        _mapView.onCreate(_mapViewBundle);
+        _mapView.getMapAsync(this);
+        return view;
+    }
+
+    /**
+     * If needed, retrieves the user's current location. Makes use of
+     * Firestore callback interface to prevent issues from asynchronous execution
+     * @param callback
+     */
+    private void getCurrentLocation(FirestoreCallback callback) {
+        _fusedLocationClient.getCurrentLocation(LocationRequest.PRIORITY_HIGH_ACCURACY,
+                                  null)
+                             .addOnSuccessListener(location -> {
+                                 if (location != null) {
+                                     _selectedLocation = location.getLatitude() +
+                                                         " " + location.getLongitude();
+                                     callback.onCallback(false);
+                                 }
+                                 else {
+                                     callback.onCallback(true);
+                                 }
+                             });
+    }
```

Comment:
This essentially used @rudy-patel 's callback interface to ensure that the current location is retrieved before any operations that depend on it are carried out